File: src-tauri/backend/untils/maafw.py
import re
import time
from pathlib import Path
from queue import Empty, Queue
from threading import Lock
from typing import Callable, List, Optional, Tuple, Union

from PIL import Image
from maa.agent_client import AgentClient
from maa.context import ContextEventSink
from maa.controller import AdbController, Win32Controller
from maa.define import MaaWin32ScreencapMethodEnum
from maa.event_sink import NotificationType
from maa.resource import Resource
from maa.tasker import Tasker, RecognitionDetail, TaskerEventSink
from maa.toolkit import Toolkit, AdbDevice, DesktopWindow
from numpy import ndarray


class DebugStreamBroker:
    """简易的 SSE 事件分发器"""

    def __init__(self):
        self._clients: List[Queue] = []
        self._lock = Lock()

    def register(self) -> Queue:
        q = Queue()
        with self._lock:
            self._clients.append(q)
        return q

    def unregister(self, q: Queue):
        with self._lock:
            if q in self._clients:
                self._clients.remove(q)

    def publish(self, payload: dict):
        if not payload:
            return
        with self._lock:
            for q in list(self._clients):
                try:
                    q.put_nowait(payload)
                except Exception:
                    # 忽略单个客户端队列的异常，避免阻塞其他客户端
                    pass


debug_broker = DebugStreamBroker()


class MaaFW:
    resource: Optional[Resource]
    controller: Union[AdbController, Win32Controller, None]
    tasker: Optional[Tasker]
    agent: Optional[AgentClient]
    context_sink=False
    tasker_sink=False

    def __init__(self):
        self.im = None
        Toolkit.init_option("./")
        Tasker.set_debug_mode(True)

        self.resource = None
        self.controller = None
        self.tasker = None
        self.agent = None
        self.notification_handler = None

    @staticmethod
    def detect_adb() -> List[AdbDevice]:
        return Toolkit.find_adb_devices()

    @staticmethod
    def detect_win32hwnd(window_regex: str) -> List[DesktopWindow]:
        windows = Toolkit.find_desktop_windows()
        result = []
        for win in windows:
            if not re.search(window_regex, win.window_name):
                continue

            result.append(win)

        return result

    def connect_adb(
            self, path: Path, address: str, config: dict
    ) -> Tuple[bool, Optional[str]]:
        self.controller = AdbController(path, address, config=config)
        connected = self.controller.post_connection().wait().succeeded
        if not connected:
            return (False, f"Failed to connect {path} {address}")

        return True, None


    def connect_win32hwnd(
            self, hwnd: Union[int, str], screencap_method: int =None, mouse_method: int =None,keyboard_method:int =None
    ) -> Tuple[bool, Optional[str]]:
        if isinstance(hwnd, str):
            hwnd = int(hwnd, 16)

        self.controller = Win32Controller(
            hwnd, screencap_method=screencap_method, mouse_method=mouse_method,keyboard_method=keyboard_method
        )
        connected = self.controller.post_connection().wait().succeeded
        if not connected:
            return (False, f"Failed to connect {hex(hwnd)}")

        return True, None

    def load_resource(self, dir: List[Path]) -> Tuple[bool, Optional[str]]:
        if not self.resource:
            self.resource = Resource()
        dir = [Path(p) for p in dir]
        for d in dir:
            if not d.exists():
                return (False, f"{d} does not exist.")

            status = self.resource.post_bundle(d).wait().succeeded
            if not status:
                return (
                    False,
                    "Fail to load resource,please check the outputs of CLI.",
                )
        return True, None

    def create_agent(self, identifier: str) -> str:
        if not self.resource:
            self.resource = Resource()

        self.agent = AgentClient(identifier)
        self.agent.bind(self.resource)
        if not self.agent.identifier:
            raise RuntimeError("Failed to create agent")

        return self.agent.identifier

    def connect_agent(self) -> Tuple[bool, Optional[str]]:
        ret = self.agent.connect()
        if not ret:
            return (None, "Failed to connect agent")
        return True, None

    def disconnect_adb(self):
        if self.controller:
            self.controller=None
            self.tasker.controller=None
        return True

    def run_task(
            self, entry: str, pipeline_override: dict = {}
    ):

        if not self.tasker:
            self.tasker = Tasker()

        if not self.resource or not self.controller:
            return (False, "Resource or Controller not initialized")

        self.tasker.bind(self.resource, self.controller)
        if not self.tasker.inited:
            return (False, "Failed to init MaaFramework tasker")
        if not self.context_sink:
            self.tasker.add_context_sink(MyNotificationHandler(debug_broker))
            self.context_sink=True
        if not self.tasker_sink:
            self.tasker.add_sink(NotificationHandler(debug_broker))
            self.tasker_sink=True
        self.tasker.post_task(entry, pipeline_override)

        return None
    def run_re(self):

        if not self.tasker:
            self.tasker = Tasker()

        if not self.resource or not self.controller:
            return (False, "Resource or Controller not initialized")

        self.tasker.bind(self.resource, self.controller)
        if not self.tasker.inited:
            return (False, "Failed to init MaaFramework tasker")
        if not self.context_sink:
            self.tasker.add_context_sink(MyNotificationHandler(debug_broker))
            self.context_sink=True
        if not self.tasker_sink:
            self.tasker.add_sink(NotificationHandler(debug_broker))
            self.tasker_sink=True

        return self.tasker

    def stop_task(self):
        if not self.tasker:
            return

        self.tasker.post_stop().wait()

    def screencap(self, capture: bool = True) -> Optional[Image.Image]:
        if not self.controller:
            return None

        if capture:
            self.im=self.controller.post_screencap().wait().get()
        # self.im = self.controller.cached_image
        if self.im is None:
            return None
        return cvmat_to_image(self.im)

    def click(self, x, y) -> bool:
        if not self.controller:
            return False

        return self.controller.post_click(x, y).wait().succeeded

    def get_reco_detail(self, reco_id: int) -> Optional[RecognitionDetail]:
        """根据 reco_id 获取识别详情，并预先将绘制结果转换为 PIL Image"""
        if not self.tasker:
            return None
        detail = self.tasker.get_recognition_detail(reco_id)
        if not detail:
            return None
        try:
            if getattr(detail, "raw_image", None) is not None and not isinstance(detail.raw_image, Image.Image):
                detail.raw_image = cvmat_to_image(detail.raw_image)
            if getattr(detail, "draw_images", None):
                detail.draw_images = [cvmat_to_image(img) for img in detail.draw_images if img is not None]
        except Exception:
            # 转换失败不影响后续流程
            pass

        return detail

    def clear_cache(self) -> bool:
        if not self.tasker:
            return False

        return self.tasker.clear_cache()


# 单例实例，供路由直接使用
maafw = MaaFW()


class MyNotificationHandler(ContextEventSink):
    """通知处理器类，处理识别事件并透传到 SSE"""

    def __init__(self, broker: DebugStreamBroker) -> None:
        super().__init__()
        self.broker = broker

    @staticmethod
    def _normalize_next_list(next_list):
        result = []
        for item in next_list or []:
            result.append(
                {
                    "name": getattr(item, "name", "") or "",
                    "jump_back": bool(getattr(item, "jump_back", False)),
                    "anchor": bool(getattr(item, "anchor", False)),
                }
            )
        return result

    def on_node_next_list(
        self,
        context: ContextEventSink,
        noti_type: NotificationType,
        detail: ContextEventSink.NodeNextListDetail,
    ):
        if noti_type != NotificationType.Starting:
            return

        payload = {
            "type": "node_next_list",
            "task_id": detail.task_id,
            "name": detail.name,
            "next_list": self._normalize_next_list(detail.next_list),
            "focus": getattr(detail, "focus", None),
            "timestamp": int(time.time() * 1000),
        }
        print(f"开始识别:{detail}")
        self.broker.publish(payload)

    def on_node_recognition(
        self,
        context: ContextEventSink,
        noti_type: NotificationType,
        detail: ContextEventSink.NodeRecognitionDetail,
    ):
        status_map = {
            NotificationType.Starting: "starting",
            NotificationType.Succeeded: "succeeded",
            NotificationType.Failed: "failed",
        }
        status = status_map.get(noti_type, "unknown")

        payload = {
            "type": "node_recognition",
            "task_id": detail.task_id,
            "reco_id": detail.reco_id,
            "name": detail.name,
            "status": status,
            "focus": getattr(detail, "focus", None),
            "timestamp": int(time.time() * 1000),
        }

        self.broker.publish(payload)

class NotificationHandler(TaskerEventSink):
    def __init__(self, broker: DebugStreamBroker) -> None:
        super().__init__()
        self.broker = broker

    def on_tasker_task(self, tasker: Tasker, noti_type: NotificationType, detail: TaskerEventSink.TaskerTaskDetail):
        pass


def cvmat_to_image(cvmat: ndarray) -> Image.Image:
    pil = Image.fromarray(cvmat)
    b, g, r = pil.split()
    return Image.merge("RGB", (r, g, b))

maafw = MaaFW()

File: src-tauri/src/maafw.rs
//! MaaFramework wrapper - equivalent to Python MaaFW class
//!
//! This module wraps the maa-framework-rs API to provide the same interface
//! as the original Python implementation.

use crate::config::DeviceInfo;
use crate::events::DebugEventBroker;
use crate::response::{BoxScore, RecognitionDetail};
use maa_framework::controller::Controller;
use maa_framework::resource::Resource;
use maa_framework::tasker::Tasker;
use maa_framework::toolkit::Toolkit;
use maa_framework::{set_debug_mode, sys};
use std::sync::Arc;
use std::time::{Duration, Instant};

/// Async wait for controller operation with timeout using spawn_blocking
pub async fn wait_with_timeout_async(controller: &Controller, id: sys::MaaId, timeout_ms: u64) -> bool {
    let ctrl_clone = controller.clone();
    tokio::task::spawn_blocking(move || {
        wait_with_timeout(&ctrl_clone, id, timeout_ms)
    }).await.unwrap_or(false)
}

/// Async wait for connection with longer timeout
#[allow(dead_code)]
pub async fn wait_for_connection_async(controller: &Controller, id: sys::MaaId) -> bool {
    wait_with_timeout_async(controller, id, CONNECTION_TIMEOUT_MS).await
}

/// Async wait with default timeout
pub async fn wait_with_default_timeout_async(controller: &Controller, id: sys::MaaId) -> bool {
    wait_with_timeout_async(controller, id, CONTROLLER_TIMEOUT_MS).await
}

/// Timeout for controller operations (screencap, connection, etc.)
const CONTROLLER_TIMEOUT_MS: u64 = 30000;
/// Timeout for connection operations (longer for emulators)
const CONNECTION_TIMEOUT_MS: u64 = 60000;
/// Poll interval for checking operation status
const STATUS_POLL_INTERVAL_MS: u64 = 50;

/// Wait for controller operation with timeout
/// Returns true if operation succeeded, false if failed or timed out
fn wait_with_timeout(controller: &Controller, id: sys::MaaId, timeout_ms: u64) -> bool {
    let start = Instant::now();
    let timeout = Duration::from_millis(timeout_ms);
    let poll_interval = Duration::from_millis(STATUS_POLL_INTERVAL_MS);

    loop {
        let status = controller.status(id);

        if status.succeeded() {
            return true;
        }
        if status.failed() {
            eprintln!("Controller operation failed");
            return false;
        }

        if start.elapsed() > timeout {
            eprintln!(
                "Controller operation timeout after {}ms",
                timeout_ms
            );
            return false;
        }

        std::thread::sleep(poll_interval);
    }
}

/// Wait for controller operation with default timeout
#[allow(dead_code)]
fn wait_with_default_timeout(controller: &Controller, id: sys::MaaId) -> bool {
    wait_with_timeout(controller, id, CONTROLLER_TIMEOUT_MS)
}

/// Wait for connection operation with longer timeout
fn wait_for_connection(controller: &Controller, id: sys::MaaId) -> bool {
    wait_with_timeout(controller, id, CONNECTION_TIMEOUT_MS)
}

/// MaaFramework wrapper - equivalent to Python MaaFW class
pub struct MaaFrameworkWrapper {
    resource: Option<Resource>,
    controller: Option<Controller>,
    tasker: Option<Tasker>,
    event_broker: Option<Arc<DebugEventBroker>>,
}

impl MaaFrameworkWrapper {
    pub fn new() -> Self {
        let target_dir = std::env::current_dir()
            .unwrap_or_else(|_| std::path::PathBuf::from("."))
            .join("target");
        let _ = Toolkit::init_option(&target_dir.to_string_lossy(), "{}");

        let _ = set_debug_mode(true);

        Self {
            resource: None,
            controller: None,
            tasker: None,
            event_broker: None,
        }
    }

    /// Find ADB devices using Toolkit
    pub fn find_adb_devices() -> Vec<DeviceInfo> {
        match Toolkit::find_adb_devices() {
            Ok(devices) => devices
                .iter()
                .map(|d| DeviceInfo {
                    name: Some(d.name.clone()),
                    device_type: Some("adb".to_string()),
                    address: Some(d.address.clone()),
                    adb_path: d.adb_path.to_str().map(|s| s.to_string()),
                    config: Some(d.config.clone()),
                    ..Default::default()
                })
                .collect(),
            Err(e) => {
                eprintln!("Failed to find ADB devices: {}", e);
                Vec::new()
            }
        }
    }

    /// Find desktop windows (Win32) using Toolkit
    pub fn find_desktop_windows() -> Vec<DeviceInfo> {
        match Toolkit::find_desktop_windows() {
            Ok(windows) => windows
                .iter()
                .map(|w| DeviceInfo {
                    name: Some(w.window_name.clone()),
                    device_type: Some("win32control".to_string()),
                    hwnd: Some(w.hwnd as i64),
                    class_name: Some(w.class_name.clone()),
                    window_name: Some(w.window_name.clone()),
                    ..Default::default()
                })
                .collect(),
            Err(e) => {
                eprintln!("Failed to find desktop windows: {}", e);
                Vec::new()
            }
        }
    }

    /// Connect to ADB device (async version)
    pub async fn connect_adb_async(
        &mut self,
        adb_path: &str,
        address: &str,
        config: serde_json::Value,
    ) -> (bool, Option<String>) {
        let config_str = serde_json::to_string(&config).unwrap_or_else(|_| "{}".to_string());
        let adb_path_str = adb_path.to_string();
        let address_str = address.to_string();

        let result = tokio::task::spawn_blocking(move || {
            match Controller::new_adb(&adb_path_str, &address_str, &config_str, "") {
                Ok(ctrl) => match ctrl.post_connection() {
                    Ok(id) => {
                        let success = wait_for_connection(&ctrl, id);
                        if success {
                            Some(ctrl)
                        } else {
                            None
                        }
                    }
                    Err(_) => None,
                },
                Err(_) => None,
            }
        }).await.unwrap_or(None);

        match result {
            Some(ctrl) => {
                self.controller = Some(ctrl);
                (true, None)
            }
            None => (false, Some(format!("Connection timeout or failed for {}", address)))
        }
    }

    /// Connect to Win32 window (async version)
    pub async fn connect_win32_async(
        &mut self,
        hwnd: i64,
        screencap_method: Option<i32>,
        mouse_method: Option<i32>,
        keyboard_method: Option<i32>,
    ) -> (bool, Option<String>) {
        let screencap = screencap_method.unwrap_or(sys::MaaWin32ScreencapMethod_GDI as i32) as sys::MaaWin32ScreencapMethod;
        let mouse = mouse_method.unwrap_or(sys::MaaWin32InputMethod_SendMessage as i32) as sys::MaaWin32InputMethod;
        let keyboard = keyboard_method.unwrap_or(sys::MaaWin32InputMethod_SendMessage as i32) as sys::MaaWin32InputMethod;

        let result = tokio::task::spawn_blocking(move || {
            match Controller::new_win32(hwnd as *mut std::ffi::c_void, screencap, mouse, keyboard) {
                Ok(ctrl) => {
                    match ctrl.post_connection() {
                        Ok(id) => {
                            let success = wait_for_connection(&ctrl, id);
                            if success {
                                Some(ctrl)
                            } else {
                                None
                            }
                        }
                        Err(_) => None,
                    }
                }
                Err(_) => None,
            }
        }).await.unwrap_or(None);

        match result {
            Some(ctrl) => {
                self.controller = Some(ctrl);
                (true, None)
            }
            None => (false, Some(format!("Connection timeout or failed for hwnd {}", hwnd)))
        }
    }

    /// Run task with pipeline override
    pub fn run_task(
        &mut self,
        entry: &str,
        pipeline_override: serde_json::Value,
    ) -> Option<String> {
        if self.tasker.is_none() {
            match Tasker::new() {
                Ok(t) => self.tasker = Some(t),
                Err(e) => return Some(format!("Failed to create tasker: {}", e)),
            }
        }

        let tasker = self.tasker.as_ref().unwrap();

        if self.resource.is_none() {
            return Some("Resource not initialized".to_string());
        }
        if self.controller.is_none() {
            return Some("Controller not initialized".to_string());
        }

        let resource = self.resource.as_ref().unwrap();
        let controller = self.controller.as_ref().unwrap();

        if let Err(e) = tasker.bind(resource, controller) {
            return Some(format!("Failed to bind resource and controller: {}", e));
        }

        if !tasker.inited() {
            return Some("Failed to init MaaFramework tasker".to_string());
        }

        if let Some(ref broker) = self.event_broker {
            let broker_clone = Arc::clone(broker);

            let _ = tasker.add_context_sink(move |msg, details| {
                let noti_type = if msg.ends_with(".Starting") {
                    "starting"
                } else if msg.ends_with(".Succeeded") {
                    "succeeded"
                } else if msg.ends_with(".Failed") {
                    "failed"
                } else {
                    "unknown"
                };

                let detail: serde_json::Value =
                    serde_json::from_str(details).unwrap_or(serde_json::Value::Null);

                if msg.starts_with("Node.NextList") {
                    let task_id =
                        detail.get("task_id").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
                    let name = detail
                        .get("name")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();

                    let next_list: Vec<(String, bool, bool)> = detail
                        .get("list")
                        .and_then(|v| v.as_array())
                        .map(|arr| {
                            arr.iter()
                                .filter_map(|item| {
                                    let name =
                                        item.get("name").and_then(|v| v.as_str())?.to_string();
                                    let jump_back = item
                                        .get("jump_back")
                                        .and_then(|v| v.as_bool())
                                        .unwrap_or(false);
                                    let anchor = item
                                        .get("anchor")
                                        .and_then(|v| v.as_bool())
                                        .unwrap_or(false);
                                    Some((name, jump_back, anchor))
                                })
                                .collect()
                        })
                        .unwrap_or_default();

                    let focus = detail.get("focus").cloned();
                    broker_clone.emit_node_next_list(task_id, name, next_list, focus);
                } else if msg.starts_with("Node.Recognition") {
                    let task_id =
                        detail.get("task_id").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
                    let reco_id =
                        detail.get("reco_id").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
                    let name = detail
                        .get("name")
                        .and_then(|v| v.as_str())
                        .unwrap_or("")
                        .to_string();
                    let focus = detail.get("focus").cloned();

                    broker_clone.emit_node_recognition(
                        task_id,
                        reco_id,
                        name,
                        noti_type.to_string(),
                        focus,
                    );
                }
            });
        }

        match tasker.post_task_json(entry, &pipeline_override) {
            Ok(_job) => None,
            Err(e) => Some(format!("Task execution error: {}", e)),
        }
    }

    /// Stop task
    pub fn stop_task(&mut self) {
        if let Some(tasker) = &self.tasker {
            let _ = tasker.post_stop();
        }
    }

    /// Check if tasker is running
    pub fn is_running(&self) -> bool {
        self.tasker.as_ref().map(|t| t.running()).unwrap_or(false)
    }

    /// Take screenshot and return as base64 with image size (async version)
    pub async fn screencap_async(&mut self) -> Option<(String, Vec<i32>)> {
        let controller = self.controller.as_ref()?;
        let ctrl_clone = controller.clone();

        let id = ctrl_clone.post_screencap().ok()?;
        
        let success = wait_with_default_timeout_async(&ctrl_clone, id).await;
        if !success {
            eprintln!("Screencap timeout or failed");
            return None;
        }

        match ctrl_clone.cached_image() {
            Ok(img_buffer) => {
                match img_buffer.to_vec() {
                    Some(raw_data) => {
                        self.encode_image_as_base64(&raw_data).map(|image| {
                            let size = self.detect_image_size(&raw_data).unwrap_or(vec![1280, 720]);
                            (image, size)
                        })
                    }
                    None => None,
                }
            }
            Err(e) => {
                eprintln!("Failed to get cached image: {}", e);
                None
            }
        }
    }

    /// Encode raw image data as base64 PNG
    fn encode_image_as_base64(&self, raw_data: &[u8]) -> Option<String> {
        if raw_data.is_empty() {
            return None;
        }

        let is_png = raw_data.len() > 4 && raw_data[0..4] == [0x89, 0x50, 0x4E, 0x47];
        let is_jpeg = raw_data.len() > 3 && raw_data[0..3] == [0xFF, 0xD8, 0xFF];

        let mime = if is_png {
            "image/png"
        } else if is_jpeg {
            "image/jpeg"
        } else {
            "image/png"
        };

        let encoded = base64::Engine::encode(&base64::engine::general_purpose::STANDARD, raw_data);
        Some(format!("data:{};base64,{}", mime, encoded))
    }

    fn detect_image_size(&self, raw_data: &[u8]) -> Option<Vec<i32>> {
        let image = image::load_from_memory(raw_data).ok()?;
        Some(vec![image.width() as i32, image.height() as i32])
    }

    /// Get recognition detail by reco_id
    pub fn get_reco_detail(&self, reco_id: i32) -> Option<RecognitionDetail> {
        let tasker = self.tasker.as_ref()?;

        match tasker.get_recognition_detail(reco_id as sys::MaaId) {
            Ok(Some(detail)) => {
                Some(RecognitionDetail {
                    reco_id: Some(reco_id),
                    name: Some(detail.node_name.clone()),
                    algorithm: Some(format!("{:?}", detail.algorithm)),
                    hit: detail.hit,
                    bbox: Some(vec![
                        detail.box_rect.x,
                        detail.box_rect.y,
                        detail.box_rect.width,
                        detail.box_rect.height,
                    ]),
                    all_results: Some(
                        detail
                            .sub_details
                            .iter()
                            .map(|sub| BoxScore {
                                bbox: Some(vec![
                                    sub.box_rect.x,
                                    sub.box_rect.y,
                                    sub.box_rect.width,
                                    sub.box_rect.height,
                                ]),
                                score: None,
                            })
                            .collect(),
                    ),
                    filtered_results: None,
                    best_result: if detail.hit {
                        Some(BoxScore {
                            bbox: Some(vec![
                                detail.box_rect.x,
                                detail.box_rect.y,
                                detail.box_rect.width,
                                detail.box_rect.height,
                            ]),
                            score: None,
                        })
                    } else {
                        None
                    },
                    raw_detail: Some(detail.detail.clone()),
                    raw_image: detail
                        .raw_image
                        .as_ref()
                        .and_then(|v| self.encode_raw_image(v)),
                    draw_images: Some(
                        detail
                            .draw_images
                            .iter()
                            .filter_map(|img| self.encode_raw_image(img))
                            .collect(),
                    ),
                })
            }
            _ => None,
        }
    }

    /// Encode raw image bytes as base64
    fn encode_raw_image(&self, raw_data: &[u8]) -> Option<String> {
        if raw_data.is_empty() {
            return None;
        }

        match image::load_from_memory(raw_data) {
            Ok(img) => {
                let rgb_img = img.to_rgb8();

                let mut png_data = Vec::new();
                match rgb_img.write_to(
                    &mut std::io::Cursor::new(&mut png_data),
                    image::ImageFormat::Png,
                ) {
                    Ok(_) => {
                        let encoded = base64::Engine::encode(
                            &base64::engine::general_purpose::STANDARD,
                            &png_data,
                        );
                        Some(format!("data:image/png;base64,{}", encoded))
                    }
                    Err(_) => None,
                }
            }
            Err(_) => {
                let encoded =
                    base64::Engine::encode(&base64::engine::general_purpose::STANDARD, raw_data);
                Some(format!("data:image/png;base64,{}", encoded))
            }
        }
    }

    /// Set event broker for debug events
    #[allow(dead_code)]
    pub fn set_event_broker(&mut self, broker: Arc<DebugEventBroker>) {
        self.event_broker = Some(broker);
    }

    /// Load resource from paths (async version)
    pub async fn load_resource_async(&mut self, paths: &[String]) -> (bool, Option<String>) {
        if paths.is_empty() {
            return (false, Some("No resource paths provided".to_string()));
        }

        let resource = match self.resource.take() {
            Some(existing) => existing,
            None => match Resource::new() {
                Ok(res) => res,
                Err(e) => return (false, Some(format!("Failed to create resource: {}", e))),
            },
        };

        let paths_clone = paths.to_vec();
        let result = tokio::task::spawn_blocking(move || {
            let mut load_success = true;
            let mut errors = Vec::new();

            for path in &paths_clone {
                if path.is_empty() {
                    continue;
                }
                match resource.post_bundle(path) {
                    Ok(job) => {
                        job.wait();
                    }
                    Err(e) => {
                        errors.push(format!("Failed to load '{}': {}", path, e));
                        load_success = false;
                    }
                }
            }

            (resource, load_success, errors)
        }).await;

        let (resource, load_success, errors) = match result {
            Ok(r) => r,
            Err(e) => return (false, Some(format!("Resource loading task failed: {}", e))),
        };

        if resource.loaded() {
            self.resource = Some(resource);
            if load_success {
                (true, Some(format!("Loaded {} resource path(s)", paths.len())))
            } else {
                (true, Some(format!("Partially loaded: {}", errors.join("; "))))
            }
        } else {
            self.resource = None;
            let msg = if errors.is_empty() {
                "Failed to load any resources".to_string()
            } else {
                errors.join("; ")
            };
            (false, Some(msg))
        }
    }
}

impl Default for MaaFrameworkWrapper {
    fn default() -> Self {
        Self::new()
    }
}
